ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala

```diff
@@ -48,32 +53,130 @@ private class PackageService(reloadPackageStoreIfChanged: PackageService.ReloadP
   @volatile private var state: State =
     State(Set.empty, TemplateIdMap.Empty, Map.empty, Map.empty, Map.empty)
 
-  // synchronized, so two threads cannot reload it concurrently
-  def reload(implicit
-      ec: ExecutionContext,
-      lc: LoggingContextOf[InstanceUUID],
-  ): Future[Error \/ Unit] =
-    synchronized {
-      reloadPackageStoreIfChanged(state.packageIds).map {
-        _.map {
-          case Some(diff) =>
-            this.state = this.state.append(diff)
-            logger.info(s"new package IDs loaded: ${diff.keySet.mkString(", ")}")
-            logger.debug(s"loaded diff: $diff")
-            ()
-          case None =>
-            logger.debug(s"new package IDs not found")
-            ()
+  // volatile, because we want to reduce the probability of unnecessary package updates
+  // due to multiple threads reading this.
+  // Writes to this are synchronized with a mutex, so in theory this should work out fine.
+  @volatile private var lastUpdated = LocalDateTime.MIN
+
+  // Regular updates should happen regardless of the current state every minute.
+  private def packagesShouldBeFetchedAgain =
+    lastUpdated.until(LocalDateTime.now(), temporal.ChronoUnit.SECONDS) >= 60L
+
+  private val _reload = {
+    // Because things can go so easily wrong, the declaration of the mutex
+    // was moved into a separate block, so it's easier to track down all interactions.
+    val packageUpdateSync = util.Concurrent.Mutex()
+    val updateTaskUpdateSync = util.Concurrent.Mutex()
+    @volatile var ongoingPackageUpdateTask = Future.successful(().right)
+    @volatile var isValid = false
+
+    def reload(jwt: Jwt)(implicit
+        ec: ExecutionContext,
+        lc: LoggingContextOf[InstanceUUID],
+    ): Future[Error \/ Unit] = {
```

Comment:
I've tried to go through this method a couple of times but it's still unclear to me why:
1. we need two mutexes
2. we need explicit mutexes

Can you guide through these changes, please?

Reply:
My comment unfortunately was lost ._.

So in short.

Under the effort of preventing unnecessary concurrent package reloads, I synchronized the access to the cache with the first mutex `packageUpdateSync` with threads that cannot acquire it picking up the ongoing update task/future. The second future is there to synchronize access to that future, because it could already be outdated (I had to introduce this after the  integration tests failed, they made me aware of that problem).

I'm not super convinced by my solution for the `ongoingPackageUpdateTask` as it's forcing sequential processing. Moreover I think it would be best to have used a fibre with support for semaphores, because os thread level semaphores have quite the overhead.

PS: We can throw the whole synchronization out if we think that the reloading the package cache is super cheap. However be aware of the scenario that with over a minute since the last update, every incoming request at this time could trigger a package reload potentially causing e.g. 1000 reloads. That could result in bad performance.

Reply:
Another avenue would be a optimistic approach where we are fine with some unnecessary updates at the same time.

Reply:
I changed that stuff to use an atomic boolean and a countdownlatch